### PR TITLE
[Message actions] New UI for replies

### LINKED ISSUE #485: [Message actions] Update composer & messages UI for replies
## Changes

Follow the [designs in Figma](https://www.figma.com/file/0MMNu7cTOzLOlWb7ctTkv3/Element-X?type=design&node-id=4367%3A168017&t=ZYzKB8ZPEHPfZTXN-1).

- Display a 'preview' of the message to reply to in the composer.
- Display a similar preview in messages that reply to another message.

### PR DESCRIPTION
**Pull request by jmartinesp**
## Changes

- Add 'reply to' message summaries to both the composer and the timeline items.
- Set min size for images in the timeline.
- Create separate layouts for items with only the media item (media items replying to another message use the regular layout).
- Add `EqualWidthColumn` for chat bubble contents.
- Move `TimelineItem___Row` views to separate files since `TimelineView` was growing too much.
- Split previews for `TextComposer`.

## Why:

Closes #485 .

**Note:** based on #526 , so that PR needs to be merged first.

## Tests

Open a chat room and send some messages and media items (images, videos, files), then reply to those with both messages and media items.

**Note**: for some reason, it's not possible to reply with media to a message in the Rust SDK, so we can't do this in our apps yet. You can test that by using Element Web.

### CODE REVIEW

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/TimelineView.kt
Comment (jmartinesp) [inline features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/TimelineView.kt:90]: I'm not sure about what we should do here, to be honest. Should we block the UI until we've paginated back to that item? What if we somehow reference a non-existing one?

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/components/event/TimelineItemImageView.kt
Comment (jmartinesp) [inline features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/components/event/TimelineItemImageView.kt:39]: Do we have somewhere where we add these kind of constants?

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/utils/messagesummary/MessageSummaryFormatter.kt
Comment (jmartinesp) [inline features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/utils/messagesummary/MessageSummaryFormatter.kt:22]: This references an `impl` only class, so I had to add it to the `impl` module.

File: libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/theme/ColorsDark.kt
Comment (jmartinesp) [inline libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/theme/ColorsDark.kt:49]: It's the same color in both themes.

File: libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/timeline/MatrixTimelineItemMapper.kt
Comment (jmartinesp) [inline libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/timeline/MatrixTimelineItemMapper.kt:42]: Maybe it's not the best point to do this? It's where iOS fetches this data, but maybe we can do this in `MatrixClient` instead.

**Comment by jmartinesp** [comment]
Since I made some changes to add extra mappings to timeline events and message contents, maybe it would be nice to have an experienced second pair of eyes taking a look at it: can you do a quick review @ganfra ?

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/TimelineView.kt
Comment (ganfra) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/TimelineView.kt:90]: We should be able to call the context API somehow with the rust-sdk, so it'll fail if the event doesn't exist (which should not happen I guess?)

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/components/event/TimelineItemImageView.kt
Comment (ganfra) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/components/event/TimelineItemImageView.kt:39]: Not for now, but we could have dimension param in the ElementTheme?

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/TimelineView.kt
Comment (jmartinesp) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/TimelineView.kt:90]: We're having a kick off meeting with iOS and design today so I'll ask about how this should work both UX and logic wise.

Comment (jmartinesp) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/TimelineView.kt:90]: Apparently we're removing this logic altogether for now since we can't just back-paginate indefinitely.

Comment (ganfra) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/TimelineView.kt:90]: But like I said, the `context` api should return the event and some of the prev and next events. It's what is used for permalink too. But probably not available yet in the rust sdk?

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/components/TimelineItemEventRow.kt
Comment (ganfra) [inline features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/components/TimelineItemEventRow.kt:250]: Should we not avoid using `EqualWidthColumn` if the inReply is null?

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/TimelineView.kt
Comment (jmartinesp) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/TimelineView.kt:90]: Yes, I think SlidingSync can only handle back pagination right now and this would be like jumping in time. We'd probably have to add some kind of "load next/load previous" marks between the latest items and those retrieved with the API.

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/components/TimelineItemEventRow.kt
Comment (jmartinesp) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/components/TimelineItemEventRow.kt:250]: I just changed `EqualWidthColumn` so it doesn't re-measure everything if all the items already have the same width. If we only have 1 child (this case) that should already be the case and it will just work like a simple Column.

File: libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/BlurHashAsyncImage.kt
Comment (julioromano) [inline libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/BlurHashAsyncImage.kt:40]: This can be simplified down to something like:

```
@Composable
fun BlurHashAsyncImage(
    model: Any?,
    blurHash: String?,
    contentDescription: String?,
    modifier: Modifier = Modifier,
    contentScale: ContentScale = ContentScale.Fit,
) {
    AsyncImage(
        model = ImageRequest.Builder(LocalContext.current)
            .data(model)
            .crossfade(true)
            .build(),
        contentDescription = contentDescription,
        modifier = modifier,
        placeholder = blurHash?.let { hash ->
            BlurHash.decode(hash, 10, 10)?.asImageBitmap()?.let(::BitmapPainter)
        },
        contentScale = contentScale,
    )
}
```

Previews also work with this impl so we can add one.
What d'you think?

Or perhaps do you think this is too trivial and we can just create an utility function to build the Painter from a blurhash string?
Something like:
```
fun blurHashPainter(blurHash: String?): Painter? = blurHash?.let { hash ->
    BlurHash.decode(hash, 10, 10)?.asImageBitmap()?.let(::BitmapPainter)
}
```
And then use it with Coil's AsyncImage?

**Review by julioromano — COMMENTED** [review]
Just some suggestions wherever I could be hopefully useful

File: libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/BlurHashAsyncImage.kt
Comment (jmartinesp) [reply libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/BlurHashAsyncImage.kt:40]: Actually, I just moved this class to the design system module from a feature one since it's needed in several places. I don't mind simplifying the implementation if it has a similar UX.

Comment (julioromano) [reply libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/BlurHashAsyncImage.kt:40]: > Actually, I just moved this class to the design system module from a feature one since it's needed in several places. I don't mind simplifying the implementation if it has a similar UX.

The preview looks the same, I didn't check the crossfade animation though.

Comment (julioromano) [reply libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/BlurHashAsyncImage.kt:40]: There's one difference though, the original `BlurHashAsyncImage` applies `fillMaxSize()` to its inner AsyncImage.
We could do the same here but I just thought it'd be too opinionated for being a library composable.

Comment (julioromano) [reply libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/BlurHashAsyncImage.kt:40]: Anyhow I thought this was a new class, not a move, if you prefer I can do a separate rework PR once this is merged

Comment (jmartinesp) [reply libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/BlurHashAsyncImage.kt:40]: Yeah, that seems for the best. Also, the cross fade sometimes looks kind of weird, starting from a black square.

**Comment by jmartinesp** [comment]
@julioromano  I uploaded some changes to address your comments, mostly removing unnecessary/duplicate code and resources.

**Review by julioromano — APPROVED** [review]

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/components/TimelineItemEventRow.kt
Comment (ganfra) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/components/TimelineItemEventRow.kt:250]: Ok nice, thanks!

File: libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/BlurHashAsyncImage.kt
Comment (julioromano) [reply libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/BlurHashAsyncImage.kt]: > Is the Drawable being released from memory by the placeholder api?
Nope

> Also I think the painter should be remembered?
Yepp

Good thing Jorge didn't listen to me ;)

Comment (julioromano) [reply libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/BlurHashAsyncImage.kt]: @ganfra Why the need of calling `Bitmap.recycle()` though? Its doc string says it shouldn't normally be called.

**Pull request merged**
